runtime/memprof.c

```diff
@@ -296,20 +318,125 @@ void caml_memprof_handle_postponed() {
+    set_trigger(caml_memprof_young_trigger - n);
+  else
+    set_trigger(caml_young_alloc_start);
+}
+
+/* Renew the next sample in the minor heap. This needs to be called
+   after each minor sampling and after each minor collection. In
+   practice, because we disable sampling during callbacks, this is
+   called at each sampling (including major ones). These extra calls
+   do not change the statistical properties of the sampling because of
+   the memorylessness of the exponential distribution. */
+void caml_memprof_renew_minor_sample(void) {
+  double exp;
+  uintnat max, exp_int;
+
+#define NO_SAMPLE { set_trigger(caml_young_alloc_start); return; }
+
+  if(lambda == 0) NO_SAMPLE;
+
+  exp = - lambda_inverse * logf(mt_generate_uniform());
+  max = caml_young_ptr - caml_young_alloc_start;
+  if(exp >= (double)max) NO_SAMPLE;
+
+  exp_int = (uintnat)exp;
+  /* This assertion can be false only if [max] is not exactly
+     representable as a double, which cannot happen if the size of the
+     minor heap does not reach 2^52. */
+  CAMLassert(exp_int < max);
+
+  set_trigger(caml_young_ptr - exp_int);
+
+  next_sample_frac_part = exp - exp_int;
+  CAMLassert(0 <= next_sample_frac_part && next_sample_frac_part < 1);
+}
+
+/* Called when exceeding the threshold for the next sample in the
+   minor heap, from the C code (the handling is different when called
+   from natively compiled OCaml code). */
+void caml_memprof_track_young(tag_t tag, uintnat wosize) {
+  CAMLparam0();
+  CAMLlocal2(ephe, callstack);
+  uintnat whsize = Whsize_wosize(wosize);
+  double rest;
+  int32_t occurences;
+
+  if(caml_memprof_suspended) {
+    caml_memprof_renew_minor_sample();
+    CAMLreturn0;
+  }
+
+  /* If [lambda == 0], then [caml_memprof_young_trigger] should be
+     equal to [caml_young_alloc_start].  But this function is only
+     called with [caml_young_alloc_start <= caml_young_ptr <
+     caml_memprof_young_trigger], which is contradictory. */
+  CAMLassert(lambda > 0);
+
+  /* [rest] is the size of the part of the allocated block which is
+     after the sampling point. */
+  rest = (caml_memprof_young_trigger - caml_young_ptr) - next_sample_frac_part;
+  CAMLassert(rest >= 0);
+  occurences = mt_generate_poisson(rest) + 1;
+
+  /* Restore the minor heap in a valid state and suspend sampling for
+     calling the callback.
+     We should not call the GC before this. */
+  caml_young_ptr += whsize;
+  caml_memprof_suspended = 1;
+  caml_memprof_renew_minor_sample();
+
+  callstack = capture_callstack_minor();
+  ephe = do_callback(tag, wosize, occurences, callstack, Minor);
+
+  caml_memprof_suspended = 0;
+  if (Is_exception_result(ephe)) caml_raise(Extract_exception(ephe));
+
+  /* We can now restore the minor heap in the state needed by
+     [Alloc_small_aux].
+     We should not call the GC after this. */
+  if(caml_young_ptr - whsize < caml_young_trigger) {
+    CAML_INSTR_INT ("force_minor/memprof@", 1);
+    caml_gc_dispatch();
+  }
```

Comment:
The invariants here are subtle and undocumented and it took me a long time to work them out. I'd like this logic to be simplified a bit, or at least to have the invariants documented. In particular:

  - The word sampled is not the one that memprof_young_trigger points to (because capturing the callstack and running the callback allocates), but the one below memprof_young_trigger by a number of words equal to the allocation done by the callback. (This may be below young_limit, in which case a GC occurs).
 
  - This function is only called from `caml_alloc_small_dispatch` and `caml_young_ptr` is never modified between the time that this function returns and the time that `Alloc_small` completes. (Otherwise, the value sampled would not be the value allocated)

  - The functions `caml_ephemeron_set_key` (and `do_check_key_clean`, `do_set`, `caml_realloc_ephe_ref_table`, etc.) never inspect the new key other than to check that it is a young block. (Otherwise, they would crash upon finding that the key does not have a valid tag and contains uninitialised garbage).

The API is also a bit tricky: an arbitrary amount of user code runs between the ephemeron being created and its key being set (because finalisers, signal handlers and context switches can happen in caml_gc_dispatch above), and this computation can include further memprof sampling. So, in edge cases there can be several outstanding ephemerons with unset keys, and their keys can remain unset for a long time. In particular, polling the ephemeron to see if the sampled block is still alive does not always give the right answer.

Reply:
> The API is also a bit tricky: an arbitrary amount of user code runs between the ephemeron being created and its key being set (because finalisers, signal handlers and context switches can happen in caml_gc_dispatch above), and this computation can include further memprof sampling. So, in edge cases there can be several outstanding ephemerons with unset keys, and their keys can remain unset for a long time. In particular, polling the ephemeron to see if the sampled block is still alive does not always give the right answer.

Good catch! I've "fixed" the issue by filling the ephemeron with `Val_unit` during the `caml_gc_dispatch` call, so that it never appears empty before the block is freed. Since unboxed values like `Val_unit` are never freed by ephemerons, this also ensures that whatever the user put in the ephemeron's data will not be reclaimed.

Reply:
> The invariants here are subtle and undocumented and it took me a long time to work them out. I'd like this logic to be simplified a bit, or at least to have the invariants documented.

I tried to add comments. Better now?

I do agree that this code is rather subtle (this is why I prepared first the PR with major blocks only). That said, I don't see a good way of avoiding this complexity. One way would be to "postpone" sampling as we do for major blocks, but then, contrarily to major blocks, there is no good way of making sure the callback is indeed called shortly after....